### PR TITLE
Remember the Milk auth

### PR DESCRIPTION
**Pull request by seehamrun**
Implements a RememberTheMilkAuthServiceExtension 

In order for this to work, also had to create a new callback handler for their authorization flow. 
Verified locally that the auth flow now works, but doesn't work end to end due to issues in the importer/exporter. (Debugging those and sending a separate pull request for them). 

#217

### CODE REVIEW

File: extensions/auth/portability-auth-rememberthemilk/src/main/java/org/dataportabilityproject/auth/rememberthemilk/RememberTheMilkAuthDataGenerator.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    // NOTE: callbackBaseUrl is unused. After authentication, RememberTheMilk will redirect
+    // to the callback set when we configured the application. To change this visit:
+    // https://www.rememberthemilk.com/help/contact/support/?ctx=api.update&report=1
+    URL authUrlSigned;
+    try {
+      URL authUrlUnsigned = new URL(AUTH_URL + "?perms=" + perms);
+      authUrlSigned = signatureGenerator.getSignature(authUrlUnsigned);
+    } catch (Exception e) {
+      logger.warn("Error generating Authentication URL");
+      e.printStackTrace();
```
Comment (holachuy) [inline extensions/auth/portability-auth-rememberthemilk/src/main/java/org/dataportabilityproject/auth/rememberthemilk/RememberTheMilkAuthDataGenerator.java]: user logger for the exception?

File: extensions/auth/portability-auth-rememberthemilk/src/main/java/org/dataportabilityproject/auth/rememberthemilk/RememberTheMilkAuthDataGenerator.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+  @Override
+  public AuthData generateAuthData(
+      String callbackBaseUrl, String authCode, String id, AuthData initialAuthData, String extra) {
+    // callbackbaseurl, id, initialAuthData and extra are not used.
+    try {
+      return new TokenAuthData(getToken(authCode));
+    } catch (IOException e) {
+      logger.warn("Error getting AuthToken: " + e.getMessage());
+      e.printStackTrace();
```
Comment (holachuy) [inline extensions/auth/portability-auth-rememberthemilk/src/main/java/org/dataportabilityproject/auth/rememberthemilk/RememberTheMilkAuthDataGenerator.java]: same here

File: extensions/auth/portability-auth-rememberthemilk/src/main/java/org/dataportabilityproject/auth/rememberthemilk/RememberTheMilkAuthDataGenerator.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      throw new IOException(
+          "Bad status code: " + statusCode + " error: " + response.getStatusMessage());
+    }
+
+    AuthElement authElement = xmlMapper.readValue(response.getContent(), AuthElement.class);
+
+    Preconditions.checkState(authElement.stat.equals("ok"), "state must be ok: %s", authElement);
+    Preconditions.checkState(
+        !Strings.isNullOrEmpty(authElement.auth.token), "token must not be empty", authElement);
+    logger.debug("Auth Token: " + authElement);
```
Comment (holachuy) [inline extensions/auth/portability-auth-rememberthemilk/src/main/java/org/dataportabilityproject/auth/rememberthemilk/RememberTheMilkAuthDataGenerator.java]: Let's remove this

File: portability-core/src/main/java/org/dataportabilityproject/serviceProviders/rememberTheMilk/model/RememberTheMilkResponse.java [hunk 4]
```diff
@@ -18,7 +18,7 @@
 import com.google.api.client.util.Key;
 
 /** A generic response from the remember the milk service. */
-public abstract class RememberTheMilkResponse {
+public abstract class  RememberTheMilkResponse {
```
Comment (holachuy) [inline portability-core/src/main/java/org/dataportabilityproject/serviceProviders/rememberTheMilk/model/RememberTheMilkResponse.java]: extra space

File: portability-gateway/src/main/java/org/dataportabilityproject/gateway/reference/LegacyAuthCallbackHandler.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    Preconditions.checkArgument(
+        ReferenceApiUtils.validateRequest(exchange, HttpMethods.GET, PATH + ".*"));
+    logger.debug("received request: {}", exchange.getRequestURI());
+
+    String redirect = baseUrl + "/error";
+    try {
+      redirect = handleExchange(exchange);
+    } catch (Exception e) {
+      logger.warn("Error handling request: {}", e.getMessage());
+      e.printStackTrace();
```
Comment (holachuy) [inline portability-gateway/src/main/java/org/dataportabilityproject/gateway/reference/LegacyAuthCallbackHandler.java]: remove this?

File: extensions/auth/portability-auth-rememberthemilk/src/main/java/org/dataportabilityproject/auth/rememberthemilk/RememberTheMilkAuthDataGenerator.java (hunk 1 above)
Comment (seehamrun) [reply extensions/auth/portability-auth-rememberthemilk/src/main/java/org/dataportabilityproject/auth/rememberthemilk/RememberTheMilkAuthDataGenerator.java]: Done. Just passed message alonge

**Pull request merged**